### ISSUE TITLE
Potential shot-through condition

### ISSUE THREAD

**Issue by tobbeanton** [bird-sanctuary/bluejay#187]
### Describe the issue

We are testing Bluejay for our upcoming [Crazyflie 2.1 - brushless](https://www.bitcraze.io/2023/10/say-hello-to-the-crazyflie-2-1-brushless/). As part of this we do autonomous flight testing over and over again, we jokingly call it [infinite flight test](https://www.bitcraze.io/2023/07/the-infinite-flight/). What we have noticed is that sometimes it just resets in mid air and we directly suspected the ESC. We built a small test rig where we can measure the mosfet signals as well as the battery voltage and cycle the PWM 10% and 100% every 300ms. This way we managed to capture the voltage dip and find a H-bridge shot-though condition. This usually happens within a minute using this test setup.
![image](https://github.com/bird-sanctuary/bluejay/assets/6593186/28bdc15b-fe52-4c13-b101-39abc50584d8)
As can be seen in the image Bc and Bp mosfet signals are both on for a short period of time causing the shot-through. This happens in the transition from breaking to accelerating where it looks like Bp is one PWM cycle late (or Bc early). I'm pretty sure this appens for the other phases as well but I don't have a capture of it.

We tried BLHeli_S 16.7 on which we could not detect the shot-though condition.

The full capture is attached and can be viewed using the [Salae Logic](https://www.saleae.com/pages/downloads).
[Shot-through-mosfet-channels.zip](https://github.com/bird-sanctuary/bluejay/files/14572651/Shot-through-mosfet-channels.zip)

### Bluejay version

0.19.2 & 0.20.1-RC2

### ESC variant

O_H_10

### PWM frequency

48

### DShot bitrate

300

### Bidirectional DShot

Off

### FC firmware

Crazyflie 2024.2

### Motor size

08028

**Comment by stylesuxx** [comment]
Interesting, thank you for the detailed report. Did you by any chance try with 24kHz PWM setting too?
Also have you tried increasing dead-time to 15?

cc @damosvil - do you have any input on this, or other things you want to see tested?

**Comment by tobbeanton** [comment]
We recently tried 24KHz PWM and found the same issue, but now on phase C.
![image](https://github.com/bird-sanctuary/bluejay/assets/6593186/1d4f21e6-4744-4258-b495-501a4453ec27)

And here are the settings we used for this capture
![image](https://github.com/bird-sanctuary/bluejay/assets/6593186/b85c325b-8741-4e32-b51e-e300f615f0d5)

**Comment by stylesuxx** [comment]
> Also have you tried increasing dead-time to 15?
Just in case you missed it.

**Comment by tobbeanton** [comment]
Due to the rarity of this appearing I'm guessing it is a timing thing, probably some interrupt that triggers att just the right time, causing the PWM to be updated a bit separated in time.

**Comment by tobbeanton** [comment]
> Just in case you missed it.

I did not try dead-time 15. Would be strange if this was the cause but better be safe then sorry. Will try tomorrow.

**Comment by damosvil** [comment]
It seems a problem related to setting the PCA registers, but checking Blheli_S and Bluejay source code it seems they are both setting those registers in the same place:
https://github.com/bitdump/BLHeli/blob/ef8c1a0b644c228f07a82f3d25e6d581492eaacf/BLHeli_S%20SiLabs/BLHeli_S.asm#L1492
https://github.com/bird-sanctuary/bluejay/blob/b803b8de71a0f8f12c30b8c105d1ab9a3b287d77/Bluejay.asm#L1110

It seems that in some occasions Xp starts working a PWM cycle before Xc (updating Xp and Xc are not synchronized to the PWM cycle), something that agrees with the code. What I don't understand is why you cannot reproduce the same issue in Blheli_S, because both codebases do the same.

¿Have you found any pattern to reproduce this issue? ¿How frequent is it in your hw?
¿could you alternate one of the led GPIOs before updating the PCA registers? - If you need a customized fw to do this let us know. 
¿could you check if you can also reproduce this issue with Bluejay 0.16?

**Comment by tobbeanton** [comment]
> What I don't understand is why you cannot reproduce the same issue in Blheli_S, because both codebases do the same.
Let us try longer and perhaps we can replicate it in Blheli_S too.

> could you check if you can also reproduce this issue with Bluejay 0.16
Yes we could reproduce it, and this time it happened in the middle of breaking, not in the change from breaking to accelerating.
![image](https://github.com/bird-sanctuary/bluejay/assets/6593186/aff082d4-9cd1-4e2e-9410-26e38d864b41)

> could you alternate one of the led GPIOs before updating the PCA registers and also scope it? - If you need a customized fw to do this let us know.
We will give it a try

**Comment by damosvil** [comment]
I have been talking with Alka (the creator of AM32) and he suggests that this might be a problem related to not using a gate driver like the fd6288, that implements shoot through detection. He also said that ARM MCUs do complement the PWM in hardware so it seems it is an issue related tinywhoop hardware in general that uses EFM8BBx MCUs.
![imagen](https://github.com/bird-sanctuary/bluejay/assets/2889851/e4734839-b27c-4372-b79c-479712a68a7c)

**Comment by damosvil** [comment]
What we can do for the next version is not to update the PCA registers if the PCA counter is about to expire. This way Xp and Xc will be updated synchronized with the PCA cycle.

**Comment by tobbeanton** [comment]
Another thing I was thinking about, why we are not able to replicate it in Blheli_s 16.7. It could just be a coincident that we have not manage to catch it but we have tested for ~20min and for Blujay it usually happens within 1min. Could this be related to interrupt rather then the auto-reload registers?

**Comment by damosvil** [comment]
> To me, checking for PCA counter to expire, sound like the right way to do it. Since the auto-reload registers are used there is already a "performance" hit since it can take almost a full cycle before the PCA registers are updated. And I don't think there is any other safe way to do it.
> 
> It sounds a bit challenging to implement but we are happy to test it if you know how to do it @damosvil

Ok, I will try a modification and I will let you know

**Comment by damosvil** [comment]
> Another thing I was thinking about, why we are not able to replicate it in Blheli_s 16.7. It could just be a coincident that we have not manage to catch it but we have tested for ~20min and for Blujay it usually happens within 1min. Could this be related to interrupt rather then the auto-reload registers?

I have checked Blheli_S code again and I think that they do something to avoid the issue in the pca_int isr:
https://github.com/bitdump/BLHeli/blob/ef8c1a0b644c228f07a82f3d25e6d581492eaacf/BLHeli_S%20SiLabs/BLHeli_S.asm#L1567

But I think that ISRs add additional latency so it would be better not to update the PWM registers if PCA counter is about to expire and reorder PCA register writes.

**Comment by damosvil** [comment]
I have been checking EFMBB2 reference manual and it seems it may be not so easy to control when to load Xc and Xp registers:
![imagen](https://github.com/bird-sanctuary/bluejay/assets/2889851/11161b97-c3c8-42c6-9475-4fec9f2e1a92)
I will check Blheli_S solution again.

**Comment by damosvil** [comment]
I think that a valid solution would be that, when a new dshot frame arrives, to store the power and damp values, and activate the PCA interrupt (generated when PCA counter is 0). In the interrupt we should set Xp and then Xc, so when the up edges happen both autoreload values are loaded in the same cycle, and disable the interrupt again. I will try to code this solution next week.

**Comment by tobbeanton** [comment]
> I think that a valid solution would be that, when a new dshot frame arrives, to store the power and damp values, and activate the PCA interrupt (generated when PCA counter is 0). In the interrupt we should set Xp and then Xc, so when the up edges happen both autoreload values are loaded in the same cycle, and disable the interrupt again. I will try to code this solution next week.

Sound good, I think this is a common way to handle it.

**Comment by tobbeanton** [comment]
Just checking how things are going? Anything we can do to help (but doing the actual fix might be above our skill level)?

**Comment by stylesuxx** [comment]
Hey, just a heads-up. We have not forgotten you, unfortunately we are currently a bit swamped with private life/work so things will take some time.

**Comment by stylesuxx** [comment]
@tobbeanton thank you, please let us know how it goes - if it works, we would appreciate a PR.

**Comment by hyp0dermik-code** [comment]
> > What I don't understand is why you cannot reproduce the same issue in Blheli_S, because both codebases do the same.
> 
> Let us try longer and perhaps we can replicate it in Blheli_S too.
> 
> > could you check if you can also reproduce this issue with Bluejay 0.16
> 
> Yes we could reproduce it, and this time it happened in the middle of breaking, not in the change from breaking to accelerating. ![image](https://private-user-images.githubusercontent.com/6593186/312428378-aff082d4-9cd1-4e2e-9410-26e38d864b41.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU2NjcwNTAsIm5iZiI6MTcxNTY2Njc1MCwicGF0aCI6Ii82NTkzMTg2LzMxMjQyODM3OC1hZmYwODJkNC05Y2QxLTRlMmUtOTQxMC0yNmUzOGQ4NjRiNDEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUxNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MTRUMDYwNTUwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjgzOWQ0YzY2ZWI0YTZhNzY1MWEzMGE1YmE5ZDBlYTZjNGM1ZjQ0NDc4MTkzYTZhYzJhYzhmYjVkM2FlNDFkNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.8dgJrJtAQiiATse9DHHcZFEjvwddZ2U3c1K9FDJmcT4)
> 
> > could you alternate one of the led GPIOs before updating the PCA registers and also scope it? - If you need a customized fw to do this let us know.
> 
> We will give it a try

This capture was from 0.16, correct?

Can you confirm what version(s) the previous 2 captures were?
https://github.com/bird-sanctuary/bluejay/issues/187#issue-2181610116
https://github.com/bird-sanctuary/bluejay/issues/187#issuecomment-1991762707

Is the timing of the bug exactly the same for each occurrence on the same version, or is there some variation? How many samples?

What variation did you see between 0.19.2 and 0.21RC (0.20.1?) Are you able to provide some instances from the missing version please?

**Comment by tobbeanton** [comment]
Is there a fix in 0.21RC? Else the bug has more or less been fully identified...?

**Comment by hyp0dermik-code** [comment]
> Is there a fix in 0.21RC? Else the bug has more or less been fully identified...?

No, I was more curious as to what the difference was in timing between 0.21 and .19.2 (if any)

**Comment by tobbeanton** [comment]
I think the bug has been there for a long time, since the PCA switching code was changed.

**Comment by alinneacsu** [comment]
Hello! Was the bug fixed in the lastest version ?

**Comment by stylesuxx** [comment]
@alinneacsu No, otherwise we would have closed the issue and mentioned in the release notes. Are you experiencing the same issues?

**Comment by alinneacsu** [comment]
@stylesuxx 

I think the issue can be similar: my setup includes a FC, based on H743, running Arducopter (bidir dshot enabled) and a 4in1 ESC running BlueJay latest version. Rarely, until now the rate is 1:50 flights, one of the motors simply turns off in flight, but it looks like it is not  demag/desync, based on logs. Tried both 24Khz / 48Khz versions, no differences.

Didn't identified any way to replicate the issue, in a controlled environment.

I have many logs indicating this situation, i'm attaching a simple screenshot for now, the constant RPM at the end of the log indicates the moment when the motor stopped:

![Screenshot 2024-11-01 at 23 28 44](https://github.com/user-attachments/assets/9ec3f1ae-1b47-41a7-a968-c7d497924700)

I'm logging also the following EDT fields:
.SS -> EDT Stress Level (120 constantly)
.SA -> EDT Status (193, rarely goes to 1)

**Comment by stylesuxx** [comment]
Please attach full logs, so people can look through them.

**Comment by alinneacsu** [comment]
Check this out: https://drive.google.com/drive/folders/1dsq6q2YpsevknT9BYpLhBDjFDhUS9mEA

**Comment by stylesuxx** [comment]
@alinneacsu can you provide some time stamps of interest for those logs please?

Also, what else have you done to troubleshoot this issue? Does it always happen with the same motor? Have you tried to change timings?

The initial issue seems to be reproducable pretty consistently at least at this one setup. So I am not sure if we are looking at the same issue here.

**Comment by Z-Vladus** [comment]
Hello everyone!
Looks like I have encountered the same consistant problem. The current decreases quickly, but still gradually and the quadcopter falls out of the sky.
Hope this bug will be fixed soon.

I may provide data if needed, just tell how and what to provide.

Is there any advice on how to avoid this issue or how to minimize it's probability?  

website log:
```
2024-12-31 @ 19:11:09 -- Done reading ESCs
2024-12-31 @ 19:11:09 -- Read ESC 4: J-H-30 - Bluejay, 0.21.0, 48kHz
2024-12-31 @ 19:11:09 -- Read ESC 3: J-H-30 - Bluejay, 0.21.0, 48kHz
2024-12-31 @ 19:11:09 -- Read ESC 2: J-H-30 - Bluejay, 0.21.0, 48kHz
2024-12-31 @ 19:11:08 -- Read ESC 1: J-H-30 - Bluejay, 0.21.0, 48kHz
2024-12-31 @ 19:11:08 -- Trying to read 4 ESCs
2024-12-31 @ 19:11:05 -- Unique device ID received - 0x01f002c3433510539363437
2024-12-31 @ 19:11:05 -- Board: S7X2, version: 0
2024-12-31 @ 19:11:05 -- Running firmware released on: Dec 26 2024 15:34:39
2024-12-31 @ 19:11:05 -- Flight controller info, identifier: BTFL version: 4.6.0
2024-12-31 @ 19:11:05 -- MultiWii API version received: 1.47.0
2024-12-31 @ 19:11:05 -- Opened serial port
2024-12-31 @ 19:11:03 -- Plugged in
2024-12-31 @ 19:11:03 -- Chrome (131.0.0.0) on Windows
```
![ESC_config](https://github.com/user-attachments/assets/4d8f2e91-f6b0-4b98-825a-c35d39ffec02)

**Comment by tobbeanton** [comment]
@Z-Vladus 

> Is there any advice on how to avoid this issue or how to minimize it's probability?

We sort of have a workaround by lowering the "maximum break strength" from 255 to 100. This reduces the problem so we don't get a reset but also impacts the motor break performance. It is the best workaround we have found so far.
